pkg/controller/config-controller.go

```diff
@@ -1,377 +1,238 @@
+	configController, err := controller.New("config-controller", mgr, controller.Options{
+		Reconciler: &CDIConfigReconciler{
+			Client:                 mgr.GetClient(),
+			Scheme:                 mgr.GetScheme(),
+			CdiClient:              cdiClient,
+			K8sClient:              k8sClient,
+			Log:                    log,
+			UploadProxyServiceName: uploadProxyServiceName,
+			ConfigName:             configName,
+			CDINamespace:           util.GetNamespace(),
+		},
+		MaxConcurrentReconciles: 2, //Equivalent to 'threadiness' in original configuration
+	})
+	if err != nil {
+		return nil, err
 	}
-
-	return nil
+	if err := addConfigControllerWatches(mgr, configController); err != nil {
+		return nil, err
+	}
+	return configController, nil
 }
 
-// Run sets up ConfigController state and executes main event loop
-func (c *ConfigController) Run(threadiness int, stopCh <-chan struct{}) error {
-	defer func() {
-		c.queue.ShutDown()
-	}()
-
-	klog.V(3).Infoln("Starting config controller Run loop")
-	if threadiness < 1 {
-		return errors.Errorf("expected >0 threads, got %d", threadiness)
+// addConfigControllerWatches sets up the watches used by the config controller.
+func addConfigControllerWatches(mgr manager.Manager, configController controller.Controller) error {
+	// Add schemes.
+	if err := cdiv1.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
 	}
-
-	informersNeedingSync := []cache.InformerSynced{c.ingressesSynced}
-	if isOpenshift := IsOpenshift(c.client); isOpenshift {
-		informersNeedingSync = append(informersNeedingSync, c.routesSynced)
+	if err := storagev1.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
 	}
-	if ok := cache.WaitForCacheSync(stopCh, informersNeedingSync...); !ok {
-		return errors.New("failed to wait for caches to sync")
+	if err := extensionsv1beta1.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
 	}
-
-	klog.V(3).Infoln("ConfigController cache has synced")
-
-	for i := 0; i < threadiness; i++ {
-		go wait.Until(c.runWorker, time.Second, stopCh)
+	if err := routev1.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
 	}
 
-	klog.Info("Started workers")
-	<-stopCh
-	klog.Info("Shutting down workers")
+	// Setup watches
+	if err := configController.Watch(&source.Kind{Type: &cdiv1.CDIConfig{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+	if err := configController.Watch(&source.Kind{Type: &storagev1.StorageClass{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+	if err := configController.Watch(&source.Kind{Type: &extensionsv1beta1.Ingress{}}, &handler.EnqueueRequestForObject{}); IgnoreIsNoMatchError(err) != nil {
+		return err
+	}
+	if err := configController.Watch(&source.Kind{Type: &routev1.Route{}}, &handler.EnqueueRequestForObject{}); IgnoreIsNoMatchError(err) != nil {
+		return err
+	}
```

Comment:
something to consider:  I think it would be more efficient to use `handler.EnqueueRequestsFromMapFunc` for routes/ingress/storageclasses.  It will return a `reconcile.Request` for the singleton config object.  We don't care about the resource that actually triggered the reconcile.  And by always reconciling on the same resource, we should reconcile less frequently when there are bursts of activity.  Not that we expect updates to be that frequent.

The operator uses `handler.EnqueueRequestsFromMapFunc` here but this implementation can be much simpler (don't have to look up anything):  https://github.com/kubevirt/containerized-data-importer/blob/master/pkg/operator/controller/scc.go#L39-L58 

May also want to filter out routes/ingress that do not refer to cdi-uploadproxy.

Reply:
Done